src/Data/Nat/Primality.agda

```diff
@@ -8,92 +8,186 @@
 open import Relation.Binary.PropositionalEquality
-  using (refl; cong)
+  using (_≡_; _≢_; refl; cong)
 
 private
   variable
-    n : ℕ
+    d m n o p : ℕ
+
+  recompute-nonTrivial : .{{NonTrivial n}} → NonTrivial n
+  recompute-nonTrivial {n} {{nontrivial}} =
+    Dec.recompute (nonTrivial? n) nontrivial
 
 ------------------------------------------------------------------------
 -- Definitions
+------------------------------------------------------------------------
 
--- Definition of compositeness
+-- The definitions of `Prime` and `Composite` in this module are
+-- intended to be built up as complementary pairs of `Decidable`
+-- predicates, with `Prime` given *negatively* as the universal closure
+-- of the negation of `Composite`, where this is given *positively* as
+-- an existential witnessing a non-trivial divisor, where such
+-- quantification is proxied by an explicit `record` type.
 
-Composite : ℕ → Set
-Composite 0 = ⊥
-Composite 1 = ⊥
-Composite n = ∃ λ d → 2 ≤ d × d < n × d ∣ n
+-- The definitions of the predicates `Composite` and `Prime`
+-- as the 'diagonal' instances of relations involving such bounds on
+-- the possible non-trivial divisors leads to the following
+-- positive/existential predicate as the basis for the whole development.
 
--- Definition of primality.
+------------------------------------------------------------------------
+-- Roughness
 
-Prime : ℕ → Set
-Prime 0 = ⊥
-Prime 1 = ⊥
-Prime n = ∀ {d} → 2 ≤ d → d < n → d ∤ n
+-- A number is m-rough if all its non-trivial divisors are bounded below
+-- by m.
+Rough : ℕ → Pred ℕ _
+Rough m n = ¬ (m HasNonTrivialDivisorLessThan n)
 
 ------------------------------------------------------------------------
--- Decidability
+-- Compositeness
 
-composite? : Decidable Composite
-composite? 0               = no λ()
-composite? 1               = no λ()
-composite? n@(suc (suc _)) = Dec.map′
-  (map₂ λ { (a , b , c) → (b , a , c)})
-  (map₂ λ { (a , b , c) → (b , a , c)})
-  (anyUpTo? (λ d → 2 ≤? d ×-dec d ∣? n) n)
+-- A number is composite if it has a proper non-trivial divisor.
+Composite : Pred ℕ _
+Composite n = n HasNonTrivialDivisorLessThan n
 
-prime? : Decidable Prime
-prime? 0               = no λ()
-prime? 1               = no λ()
-prime? n@(suc (suc _)) = Dec.map′
-  (λ f {d} → flip (f {d}))
-  (λ f {d} → flip (f {d}))
-  (allUpTo? (λ d → 2 ≤? d →-dec ¬? (d ∣? n)) n)
+------------------------------------------------------------------------
+-- Primality
+
+-- Prime as the complement of Composite (and hence the diagonal of Rough
+-- as defined above). The constructor `prime` takes a proof `notComposite`
+-- that NonTrivial p is not composite and thereby enforces that:
+-- * p is a fortiori NonZero and NonUnit
+-- * p is p-Rough, i.e. any proper divisor must be at least p, i.e. p itself
+record Prime (p : ℕ) : Set where
+  constructor prime
+  field
+    .{{nontrivial}} : NonTrivial p
+    notComposite    : ¬ Composite p
```

Comment:
So that's the definition I finally adopted, lifting the `Composite` definition over that of `Prime`. But I didn't move the code for the properties of `Composite` after that... easy to do, though!

Reply:
Now done.